arrows/klv/tests/test_klv_read_write.cxx

```diff
@@ -0,0 +1,437 @@
+  KLV_READ_INT( int32_t, 0, 3, 0x001122 )
+  KLV_READ_INT( int64_t, 0, 6, 0x001122334455 )
+
+  // Signed, negative - smaller than native size
+  KLV_READ_INT( int8_t, 8, 0, 0 )
+  KLV_READ_INT( int16_t, 8, 1, 0xFF88 )
+  KLV_READ_INT( int32_t, 8, 3, 0xFF8899aa )
+  KLV_READ_INT( int64_t, 8, 6, 0xFFFF8899aabbccdd )
+
+  // Unsigned - bigger than native size
+  KLV_READ_INT_ERR( uint8_t,  0, 2, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( uint16_t, 0, 3, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( uint32_t, 0, 5, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( uint64_t, 0, 9, kv::metadata_type_overflow )
+
+  // Signed - bigger than native size
+  KLV_READ_INT_ERR( int8_t,   0, 2, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( int16_t,  0, 3, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( int32_t,  0, 5, kv::metadata_type_overflow )
+  KLV_READ_INT_ERR( int64_t,  0, 9, kv::metadata_type_overflow )
+
+#undef KLV_READ_INT
+#undef KLV_READ_INT_ERR
+}
+
+// ----------------------------------------------------------------------------
+TEST ( klv, write_int )
+{
+#define KLV_WRITE_INT( T, length, value )                                    \
+  {                                                                          \
+    vec_t data( length, 0xba );                                              \
+    auto it = data.begin();                                                  \
+    klv_write_int< T >( static_cast< T >( value ), it, length );             \
+    EXPECT_EQ( it, data.begin() + length );                                  \
+    it = data.begin();                                                       \
+    EXPECT_EQ( static_cast< T >( value ), klv_read_int< T >( it, length ) ); \
+  }
+
+#define KLV_WRITE_INT_ERR( T, length, value, e )                          \
+  {                                                                       \
+    vec_t data( length, 0xba );                                           \
+    auto it = data.begin();                                               \
+    EXPECT_THROW(                                                         \
+      klv_write_int< T >( static_cast< T >( value ), it, length ); , e ); \
+  }
+
+  // Unsigned - arbitrary numbers
+  KLV_WRITE_INT( uint8_t,  1, 0x11 )
+  KLV_WRITE_INT( uint16_t, 2, 0x1122 )
+  KLV_WRITE_INT( uint32_t, 4, 0x11223344 )
+  KLV_WRITE_INT( uint64_t, 8, 0x1122334455667788 )
+
+  // Signed, positive - arbitrary numbers
+  KLV_WRITE_INT( int8_t,   1, 0x11 )
+  KLV_WRITE_INT( int16_t,  2, 0x1122 )
+  KLV_WRITE_INT( int32_t,  4, 0x11223344 )
+  KLV_WRITE_INT( int64_t,  8, 0x1122334455667788 )
+
+  // Signed, negative - arbitrary numbers
+  KLV_WRITE_INT( int8_t,   1, -0x11 )
+  KLV_WRITE_INT( int16_t,  2, -0x1122 )
+  KLV_WRITE_INT( int32_t,  4, -0x11223344 )
+  KLV_WRITE_INT( int64_t,  8, -0x1122334455667788 )
+
+  // Lowest representable value
+  KLV_WRITE_INT( uint8_t,  1, std::numeric_limits< uint8_t  >::min() )
```

Comment:
There's an internal email thread about `std::numeric_limits` not behaving in an intuitive fashion but I forget whether it's at all applicable here. I'll try to pull it up after the review.

Reply:
It looks like `std::numeric_limits` returns odd values for floating points. The correct solution for floating points is `std::numeric_limits<double>::lowest()`. But this appears irrelevant here.

Reply:
Thanks, I'll keep that in mind for the floating-point readers/writers.